swiftsimio/visualisation/projection_backends/gpu.py

```diff
@@ -0,0 +1,231 @@
+            return
+
+        if cells_spanned <= 1:
+            # Easygame, gg
+            cuda.atomic.add(
+                img, (particle_cell_x, particle_cell_y), mass * inverse_cell_area
+            )
+        else:
+            # Now we loop over the square of cells that the kernel lives in
+            for cell_x in range(
+                # Ensure that the lowest x value is 0, otherwise we'll segfault
+                max(0, particle_cell_x - cells_spanned),
+                # Ensure that the highest x value lies within the array bounds,
+                # otherwise we'll segfault (oops).
+                min(particle_cell_x + cells_spanned + 1, maximal_array_index),
+            ):
+                # The distance in x to our new favourite cell
+                # remember that our x, y are all in a box of [0, 1]
+                # calculate the distance to the cell center
+                distance_x = (float32(cell_x) + 0.5) * pixel_width
+                distance_x -= float32(x_pos)
+                distance_x_2 = distance_x * distance_x
+                for cell_y in range(
+                    max(0, particle_cell_y - cells_spanned),
+                    min(particle_cell_y + cells_spanned + 1, maximal_array_index),
+                ):
+                    distance_y = (float32(cell_y) + 0.5) * pixel_width
+                    distance_y -= float32(y_pos)
+                    distance_y_2 = distance_y * distance_y
+
+                    r = sqrt(distance_x_2 + distance_y_2)
+
+                    kernel_eval = kernel(r, kernel_width)
+
+                    cuda.atomic.add(img, (cell_x, cell_y), mass * kernel_eval)
+
+
+def scatter(x: float64, y: float64, m: float32, h: float32, res: int) -> ndarray:
+    """
+    Parallel implementation of scatter
+
+    Creates a weighted scatter plot. Computes contributions from
+    particles with positions (`x`,`y`) with smoothing lengths `h`
+    weighted by quantities `m`.
+    This ignores boundary effects.
+
+    Parameters
+    ----------
+    x : np.array[float64]
+        array of x-positions of the particles. Must be bounded by [0, 1].
+
+    y : np.array[float64]
+        array of y-positions of the particles. Must be bounded by [0, 1].
+
+    m : np.array[float32]
+        array of masses (or otherwise weights) of the particles
+
+    h : np.array[float32]
+        array of smoothing lengths of the particles
+
+    res : int
+        the number of pixels along one axis, i.e. this returns a square
+        of res * res.
+
+    Returns
+    -------
+
+    np.array[float32, float32, float32]
+        pixel grid of quantity
+
+    See Also
+    --------
+
+    scatter : Creates 2D scatter plot from SWIFT data
+
+    Notes
+    -----
+
+    Explicitly defining the types in this function allows
+    a performance improvement.
+    """
+    if not CUDA_AVAILABLE:
+        raise Exception(
+            "Unable to load the GPU function. " "Please check your module numba.cuda."
```

Comment:
Can we raise the CudaSupportError here?

Reply:
For the tests it will become a problem. We need to know in advance if the function is available or not.